inputremapper/input_event.py

```diff
@@ -62,103 +47,62 @@ class InputEvent:
     as a drop in replacement for evdev.InputEvent
     """
 
-    message_type = MessageType.selected_event
-
     sec: int
     usec: int
     type: int
     code: int
     value: int
     actions: Tuple[EventActions, ...] = ()
+    origin: Optional[str] = None
 
-    def __hash__(self):
-        return hash((self.type, self.code, self.value))
-
-    def __eq__(self, other: Any):
+    def __eq__(self, other: InputEvent | evdev.InputEvent | Tuple[int, int, int]):
+        # useful in tests
         if isinstance(other, InputEvent) or isinstance(other, evdev.InputEvent):
             return self.event_tuple == (other.type, other.code, other.value)
         if isinstance(other, tuple):
             return self.event_tuple == other
-        return False
-
-    @classmethod
-    def __get_validators__(cls):
-        """Used by pydantic and EventCombination to create InputEvent objects."""
-        yield cls.validate
-
-    @classmethod
-    def validate(cls, init_arg: InputEventValidationType) -> InputEvent:
-        """Try all the different methods, and raise an error if none succeed."""
-        if isinstance(init_arg, InputEvent):
-            return init_arg
-
-        event = None
-        validators: Sequence[Callable[..., InputEvent]] = (
-            cls.from_event,
-            cls.from_string,
-            cls.from_tuple,
-        )
-        for validator in validators:
-            try:
-                event = validator(init_arg)
-                break
-            except InputEventCreationError:
-                pass
+        raise TypeError(f"cannot compare {type(other)} with InputEvent")
 
-        if event:
-            return event
-
-        raise ValueError(f"failed to create InputEvent with {init_arg = }")
+    @property
+    def input_match_hash(self) -> Hashable:
```

Comment:
How about adding a `__hash__` function to both InputConfig and InputEvent?

Instead of using `input_match_hash` I would probably prefer a function like

```py
def is_matching_event(event: Inputevent): bool
```

in InputConfig or something.

Reply:
So that

```py
context.notify_callbacks[input_config].append(
    handler.notify
)
```

instead of 

```py
context.notify_callbacks[input_config.input_match_hash].append(
    handler.notify
)
```

Reply:
```py
context.add_notify_callback(input_config, handler.notify)
```

instead of

```py
context.notify_callbacks[input_config.input_match_hash].append(
    handler.notify
)
```

Reply:
```py
notify_callbacks = self.context.notify_callbacks.get(event.input_match_hash)
```

instead of

```py
notify_callbacks = self.context.get_notify_callbacks(event)
```

Reply:
The hashing is kind of difficult:
The InputConfig needs two different hashes.
 1. It is itself a hash, via the `__hash__` method. 
 2. It has the input_match_hash

The first one is needed for the Preset class, as it is totally okay to have multiple mappings with the same input, as long as they have different `analog_threshold` values. This is currently implemented with the `__hash__` method because than it is automatically used whenever we use a `InputCombination` as a hash.

The second one is needed in the event pipeline to match against incoming InputEvent's. There we must ignore the `analog_threshold`.

The InputEvent only has the input_match_hash, and is never used as a hash itself.

I would rather remove the `__hash__` method and implement one for the InputCombination instead (maybe this one can elegantly solve the Permutation check). This way InputConfig can not be a hash itself. _explicite is better than implicit_ especially in this case, where confusion is likely.


> ```python
> context.add_notify_callback(input_config, handler.notify)
> ```

This method would do the same thing as we already do:
```python
def add_notify_callback(self, input_config, callback):
    self.notify_callbacks[input_config.input_match_hash].append(callback)
```

Reply:
> This method would do the same thing as we already do:

What do you mean?

Reply:
I was just thinking, there is not much the `add_` and `get_notify_callback` methods would not do much at all. But maybe it makes still sense if we want to make changes to the Context in the future.

Reply:
It makes it easier to read and use, because it is not required to remember that `input_match_hash` has to be used, and no knowledge about the datastructure of the notify callbacks would be required anymore